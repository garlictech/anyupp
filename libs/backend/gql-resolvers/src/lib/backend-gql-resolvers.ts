export function backendGqlResolvers(): string {
  return 'backend-gql-resolvers';
}
