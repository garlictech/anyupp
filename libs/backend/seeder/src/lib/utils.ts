import { seededIdPrefix } from '@bgap/shared/fixtures';
import { Role } from '@bgap/domain';

const generateUnitId = (idx: number) => `${seededIdPrefix}unit_${idx}_id`;
const generateProductCategoryId = (unitIdx: number, idx: number) =>
  `${seededIdPrefix}product_category_c${unitIdx}_${idx}_id`;
const generateUnitProductId = (unitIdx: number, idx: number) =>
  `${seededIdPrefix}unit_product_c${unitIdx}_${idx}_id`;
const generateVariantId = (
  unitIdx: number,
  productId: number,
  idx: number,
  type: string,
) =>
  `${seededIdPrefix}${type}_product_variant_c${unitIdx}_p${productId}_${idx}_id`;
const generateOrderId = (idx: number) => `${seededIdPrefix}order_${idx}_id`;
const generateUserId = (idx: number) => `${seededIdPrefix}user_${idx}_id`;
const generateRoleContextId = (idx: number, role: Role) =>
  `${seededIdPrefix}role_context_${idx}_${role}_id`;
const generateAdminRoleContextId = (
  idx: number,
  role: Role,
  username: string,
) => `${seededIdPrefix}admin_role_context_${idx}_${role}_${username}_id`;

export const seedUtils = {
  generateUnitId,
  generateProductCategoryId,
  generateUnitProductId,
  generateVariantId,
  generateOrderId,
  generateUserId,
  generateRoleContextId,
  generateAdminRoleContextId,
};
