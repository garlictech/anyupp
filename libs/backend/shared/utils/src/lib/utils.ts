import * as sst from '@serverless-stack/resources';

export const getFQParamNameFromDetails = (
  stage: string,
  appName: string,
  paramName: string,
) => `/${stage}-${appName}/generated/${paramName}`;

export const getFQParamName = (app: sst.App, paramName: string) =>
  getFQParamNameFromDetails(app.stage, app.name, paramName);

export const commonBackendStack = 'common-backend-anyupp';
export const anyuppVpcName = 'AnyuppVpc';
export const anyuppFargateClusterName = 'anyupp-fargate-cluster';
export const anyuppVpcSecurityGroupParamName =
  '/common/AnyuppVpcSecurityGroupName';
export const anyuppVpcIdParamName = '/common/AnyuppVpcId';

export const getCommonVpcNameDev = () =>
  `dev-${commonBackendStack}/${anyuppVpcName}`;

export const getCommonVpcNameProd = () =>
  `prod-${commonBackendStack}/${anyuppVpcName}`;
