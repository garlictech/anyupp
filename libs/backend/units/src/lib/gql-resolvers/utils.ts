import { CrudSdk } from '@bgap/crud-gql/api';
import { DynamoDB } from 'aws-sdk';

export interface UnitsResolverDeps {
  crudSdk: CrudSdk;
  hashGenerator: (arg: string) => string;
  uuidGenerator: () => string;
  tableName: string;
  docClient: DynamoDB.DocumentClient;
}
