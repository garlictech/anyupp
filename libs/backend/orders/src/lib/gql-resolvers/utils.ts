import * as OE from 'fp-ts-rxjs/lib/ObservableEither';
import { pipe } from 'fp-ts/lib/function';
import * as CrudApi from '@bgap/crud-gql/api';
import * as R from 'ramda';
import { shareReplay } from 'rxjs/operators';
import { AxiosStatic } from 'axios';

export interface OrderResolverDeps {
  crudSdk: CrudApi.CrudSdk;
  userId: string;
  unitTableName: string;
  currentTime: () => Date;
  random: () => number;
  axiosInstance: AxiosStatic;
}

const getUnitProductHelper = R.memoizeWith(
  (_sdk: CrudApi.CrudSdk, id: string) => id,
  (sdk: CrudApi.CrudSdk, id: string) =>
    pipe(
      sdk.GetUnitProduct({ id }),
      OE.fromPredicate(
        R.complement(R.isNil),
        () => `UnitProduct cannot be found: ${id}`,
      ),
      shareReplay(1),
    ),
);

export const getUnitProduct =
  (sdk: CrudApi.CrudSdk) =>
  (id: string): OE.ObservableEither<string, CrudApi.UnitProduct> =>
    getUnitProductHelper(sdk, id);

const getGroupProductHelper = R.memoizeWith(
  (_sdk: CrudApi.CrudSdk, id: string) => id,
  (sdk: CrudApi.CrudSdk, id: string) =>
    pipe(
      sdk.GetGroupProduct({ id }),
      OE.fromPredicate(
        R.complement(R.isNil),
        () => `GroupProduct cannot be found: ${id}`,
      ),
      shareReplay(1),
    ),
);

export const getGroupProduct =
  (sdk: CrudApi.CrudSdk) =>
  (id: string): OE.ObservableEither<string, CrudApi.GroupProduct> =>
    getGroupProductHelper(sdk, id);

const getChainProductHelper = R.memoizeWith(
  (_sdk: CrudApi.CrudSdk, id: string) => id,
  (sdk: CrudApi.CrudSdk, id: string) =>
    pipe(
      sdk.GetChainProduct({ id }),
      OE.fromPredicate(
        R.complement(R.isNil),
        () => `ChainProduct cannot be found: ${id}`,
      ),
      shareReplay(1),
    ),
);

export const getChainProduct =
  (sdk: CrudApi.CrudSdk) =>
  (id: string): OE.ObservableEither<string, CrudApi.ChainProduct> =>
    getChainProductHelper(sdk, id);

export const getGroupProductOfUnitProduct =
  (sdk: CrudApi.CrudSdk) =>
  (unitProductId: string): OE.ObservableEither<string, CrudApi.GroupProduct> =>
    pipe(
      getUnitProduct(sdk)(unitProductId),
      OE.chain(unitProduct => getGroupProduct(sdk)(unitProduct.parentId)),
    );

const getGeneratedProductHelper = R.memoizeWith(
  (_sdk: CrudApi.CrudSdk, productId: string) => productId,
  (sdk: CrudApi.CrudSdk, productId: string) =>
    pipe(
      sdk.GetGeneratedProduct({ id: productId }),
      OE.fromPredicate(
        R.complement(R.isNil),
        () => `GeneratedProduct cannot be found: ${productId}`,
      ),
      shareReplay(1),
    ),
);

export const getGeneratedProduct =
  (sdk: CrudApi.CrudSdk) =>
  (productId: string): OE.ObservableEither<string, CrudApi.GeneratedProduct> =>
    getGeneratedProductHelper(sdk, productId);

export const getAllParentsOfUnitProduct =
  (sdk: CrudApi.CrudSdk) =>
  (
    unitProductId: string,
  ): OE.ObservableEither<
    string,
    {
      unitProduct: CrudApi.UnitProduct;
      groupProduct: CrudApi.GroupProduct;
      chainProduct: CrudApi.ChainProduct;
    }
  > =>
    pipe(
      getUnitProduct(sdk)(unitProductId),
      OE.chain(unitProduct =>
        pipe(
          getGroupProduct(sdk)(unitProduct.parentId),
          OE.map(groupProduct => ({
            unitProduct,
            groupProduct,
          })),
        ),
      ),
      OE.chain(state =>
        pipe(
          getChainProduct(sdk)(state.groupProduct.parentId),
          OE.map(chainProduct => ({
            ...state,
            chainProduct,
          })),
        ),
      ),
    );
