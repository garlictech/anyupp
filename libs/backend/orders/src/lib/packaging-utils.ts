import { pipe } from 'fp-ts/lib/function';
import * as R from 'ramda';

import { OrderItem } from '@bgap/domain';

// Collect all the non-null netPackagingFee-s that are deep in the order
// config sets, then sum them up
const calculatePackagingFeeOfOrderItemConfigSets = (
  items: OrderItem[],
): number =>
  pipe(
    items,
    R.map(item =>
      pipe(
        item.configSets ?? [],
        R.map(cs => cs.items),
        x => R.reject(R.isNil)(x),
        R.flatten,
        R.map(x => x.netPackagingFee),
        R.reject(R.isNil),
        R.map(x => x * item.quantity),
      ),
    ),
    R.flatten,
    R.reduce(R.add, 0),
  );

const calculatePackagingFeeOfOrderItems = (items: OrderItem[]): number =>
  pipe(
    items,
    R.map(x => x.quantity * (x.netPackagingFee || 0)),
    R.reduce(R.add, 0),
  );

export const calculatePackagingFeeOfOrder = (items: OrderItem[]): number =>
  pipe(
    items,
    R.juxt([
      calculatePackagingFeeOfOrderItems,
      calculatePackagingFeeOfOrderItemConfigSets,
    ]),
    R.sum,
  );
