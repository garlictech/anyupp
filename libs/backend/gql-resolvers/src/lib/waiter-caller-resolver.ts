/* eskint-disable @typescript-eslint/no-explicit-any */
import { oeTryCatch } from '@bgap/shared/utils';
import { MutationCallWaiterArgs, CallWaiterInput, Unit } from '@bgap/domain';
import { CrudSdk } from '@bgap/crud-gql/api';
import * as OE from 'fp-ts-rxjs/lib/ObservableEither';
import { pipe } from 'fp-ts/lib/function';
import { throwError } from 'rxjs';
import { mapTo } from 'rxjs/operators';
import * as R from 'ramda';
import { AxiosStatic } from 'axios';
import { callWaiter } from '@bgap/rkeeper-api';

export interface WaiterCallerResolverDeps {
  sdk: CrudSdk;
  currentTimeISOString: () => string;
  axiosInstance: AxiosStatic;
  uuidGenerator: () => string;
}

export const getUnit =
  (deps: WaiterCallerResolverDeps) => (args: CallWaiterInput) =>
    pipe(
      deps.sdk.GetUnit({ id: args.unitId }),
      oeTryCatch,
      OE.chain(
        OE.fromPredicate(
          unit => !!unit,
          () => 'unit_not_found',
        ),
      ),
      OE.map(unit => unit as Unit),
    );

export const waiterCallerResolver =
  (deps: WaiterCallerResolverDeps) =>
  (args: MutationCallWaiterArgs): ReturnType<CrudSdk['CallWaiter']> =>
    pipe(
      args.input,
      OE.fromPredicate(
        x => !R.isEmpty(x?.unitId),
        () => 'unit_id_required',
      ),
      OE.map(x => x as CallWaiterInput),
      OE.chain(getUnit(deps)),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.canCallWaiter,
          () => 'unit_cannot_call_waiter',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.pos?.rkeeper?.endpointUri,
          () => 'not_supported_pos',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.pos?.rkeeper?.waiterOrderId,
          () => 'rkeeper_waiter_caller_not_configured',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.externalId,
          () => 'rkeeper_external_id_not_configured',
        ),
      ),
      OE.chain(unit => callWaiter(deps)(unit, args.input as CallWaiterInput)),
      OE.getOrElse<string, any>(throwError),
      mapTo(true),
    );
