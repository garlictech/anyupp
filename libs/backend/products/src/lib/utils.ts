import { CrudSdk } from '@bgap/crud-gql/api';
import { getAllPaginatedData } from '@bgap/gql-sdk';
import { filterNullishGraphqlListWithDefault } from '@bgap/shared/utils';
import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import {
  GroupProduct,
  SearchGroupProductsQueryVariables,
  SearchUnitProductsQueryVariables,
  UnitProduct,
} from '@bgap/domain';
import { Observable } from 'rxjs';

export interface ProductResolverDeps {
  crudSdk: CrudSdk;
  unitProductTableName: string;
  chainProductTableName: string;
  groupProductTableName: string;
  docClient: DocumentClient;
}

export const getUnitIdsFromUnitProductList = (unitProducts: UnitProduct[]) =>
  unitProducts.map(unitProduct => unitProduct.unitId);

export const listGroupProductsForChainProductParent =
  (crudSdk: CrudSdk) =>
  (chainProductId: string): Observable<Array<GroupProduct>> => {
    const input: SearchGroupProductsQueryVariables = {
      filter: { parentId: { eq: chainProductId } },
    };

    return getAllPaginatedData(crudSdk.SearchGroupProducts, {
      query: input,
      options: { fetchPolicy: 'no-cache' },
    }).pipe(filterNullishGraphqlListWithDefault<GroupProduct>([]));
  };

export const listUnitProductsForGroupProductParents =
  (crudSdk: CrudSdk) =>
  (groupProductIds: string[]): Observable<Array<UnitProduct>> => {
    const input: SearchUnitProductsQueryVariables = {
      filter: { or: groupProductIds.map(x => ({ parentId: { eq: x } })) },
    };

    return getAllPaginatedData(crudSdk.SearchUnitProducts, {
      query: input,
      options: { fetchPolicy: 'no-cache' },
    }).pipe(filterNullishGraphqlListWithDefault<UnitProduct>([]));
  };
