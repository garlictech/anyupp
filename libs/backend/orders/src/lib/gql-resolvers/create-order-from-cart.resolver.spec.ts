/* eslint-disable @typescript-eslint/no-explicit-any */
import * as E from 'fp-ts/lib/Either';
import * as CrudApi from '@bgap/crud-gql/api';
import { of, throwError } from 'rxjs';
import { tap } from 'rxjs/operators';
import {
  CalculationState_UnitAdded,
  createOrderFromCart,
  getUnit,
  getGroupCurrency,
  CalculationState_GroupCurrencyAdded,
  getNextOrderNum,
} from './create-order-from-cart.resolver';
import * as anyuppBackendLib from '@bgap/anyupp-backend-lib';
import * as utils from './utils';

jest.mock('@bgap/anyupp-backend-lib', () => ({
  __esModule: true,
  // @ts-ignore
  ...jest.requireActual('@bgap/anyupp-backend-lib'),
}));

jest.mock('./utils', () => ({
  __esModule: true,
  // @ts-ignore
  ...jest.requireActual('./utils'),
}));

const createOrderFromCart_noCart = [
  of(undefined),
  throwError('CART ERROR'),
  of({ userId: 'DIFFERENT USER ID' }),
];

test.each(createOrderFromCart_noCart)(
  'createOrderFromCart should fail without a cart',
  (returnValue, done: any) => {
    const depsFixture = {
      crudSdk: {
        GetCart: jest.fn().mockReturnValue(returnValue),
      },
      userId: 'USER ID',
    } as any;

    createOrderFromCart('CART ID')(depsFixture)
      .pipe(
        tap(res => {
          expect(res).toMatchSnapshot('RESULT');
          expect(depsFixture.crudSdk.GetCart.mock.calls).toMatchSnapshot(
            'GETCART CALLS',
          );
        }),
      )
      .subscribe(done());
  },
);

const getUnitCases = [
  { cart: { unitId: 'UNIT ID' }, unit: of(undefined) },
  { cart: { unitId: 'UNIT ID' }, unit: throwError('UNIT ERROR') },
  { cart: { unitId: 'UNIT ID' }, unit: of({ id: 'UNIT ID 1' }) },
  {
    cart: { unitId: 'UNIT ID' },
    unit: of({ id: 'UNIT ID 2', isAcceptingOrders: true }),
  },
  {
    cart: { unitId: 'UNIT ID' },
    unit: of({
      id: 'UNIT ID 3',
      isAcceptingOrders: true,
      orderPaymentPolicy: CrudApi.OrderPaymentPolicy.afterpay,
    }),
  },
  {
    cart: { unitId: 'UNIT ID' },
    unit: of({
      id: 'UNIT ID 4',
      isAcceptingOrders: true,
      orderPaymentPolicy: CrudApi.OrderPaymentPolicy.prepay,
    }),
  },
  {
    cart: { unitId: 'UNIT ID', paymentMode: {} },
    unit: of({
      id: 'UNIT ID 5',
      isAcceptingOrders: true,
    }),
  },
];

test.each(getUnitCases)('getUnit cases', ({ cart, unit }, done: any) => {
  const depsFixture = {
    crudSdk: {
      GetUnit: jest.fn().mockReturnValue(unit),
    },
  } as any;

  getUnit(depsFixture)(cart as CrudApi.Cart)
    .pipe(
      tap(res => {
        expect(depsFixture.crudSdk.GetUnit.mock.calls).toMatchSnapshot(
          'GETUNIT CALLS',
        );
        expect(res).toMatchSnapshot('RESULT');
      }),
    )
    .subscribe(done());
});

const getGroupCurrencyCases = [
  {
    inputState: { cart: { id: 'CART ID 1' }, unit: { groupId: 'GROUP ID 1' } },
    groupCurrencyResult: of(undefined),
  },
  {
    inputState: { cart: { id: 'CART ID 2' }, unit: { groupId: 'GROUP ID 2' } },
    groupCurrencyResult: throwError('GROUP CURRENCY FETCH FAILURE'),
  },
  {
    inputState: { cart: { id: 'CART ID 3' }, unit: { groupId: 'GROUP ID 3' } },
    groupCurrencyResult: of('GROUP CURRENCY'),
  },
];

test.each(getGroupCurrencyCases)(
  'getGroupCurrency cases',
  ({ inputState, groupCurrencyResult }, done: any) => {
    const depsFixture = {
      crudSdk: {
        GetGroupCurrency: jest.fn().mockReturnValue(groupCurrencyResult),
      },
    } as any;

    getGroupCurrency(depsFixture)(inputState as CalculationState_UnitAdded)
      .pipe(
        tap(res => {
          expect(
            depsFixture.crudSdk.GetGroupCurrency.mock.calls,
          ).toMatchSnapshot('GETGROUPCURRENCY CALLS');
          expect(res).toMatchSnapshot('RESULT');
        }),
      )
      .subscribe(done());
  },
);

const getNextOrderNumCases = [
  {
    inputState: {
      cart: { place: { table: 'TABLE1', seat: 'SEAT1' } },
      unit: { id: 'UNIT ID 1' },
    },
    result: of(undefined),
  },
  {
    inputState: {
      cart: { place: { table: 'TABLE2', seat: 'SEAT2' } },
      unit: { id: 'UNIT ID 2' },
    },
    result: throwError('INCREMENT ORDER NUM FAILURE'),
  },
  {
    inputState: {
      cart: { place: { table: 'TABLE3', seat: 'SEAT3' } },
      unit: { id: 'UNIT ID 3' },
    },
    result: of(10),
  },
  {
    inputState: {
      cart: {},
      unit: { id: 'UNIT ID 4' },
    },
    result: of(20),
  },
];

test.each(getNextOrderNumCases)(
  'getNextOrderNum cases',
  ({ inputState, result }, done: any) => {
    const depsFixture = {
      crudSdk: {
        GetGroupCurrency: jest.fn().mockReturnValue(result),
      },
      random: () => 0.5,
      unitTableName: 'UNIT TABLE NAME',
    } as any;

    const funcSpy = jest.fn().mockReturnValue(result);
    const incrementOrderNumSpy = jest
      .spyOn(anyuppBackendLib, 'incrementOrderNum')
      .mockReturnValue(funcSpy);

    getNextOrderNum(depsFixture)(
      inputState as CalculationState_GroupCurrencyAdded,
    )
      .pipe(
        tap(res => {
          expect(incrementOrderNumSpy.mock.calls).toMatchSnapshot(
            'INCREMENTORDERNUM DEPS CALLS',
          );
          expect(funcSpy.mock.calls).toMatchSnapshot(
            'INCREMENTORDERNUM FUNC CALLS',
          );
          expect(res).toMatchSnapshot('RESULT');
        }),
      )
      .subscribe(done());
  },
);

const getAllParentsOfUnitProductCases = [
  {
    inputState: {},
    getAllParentsOfUnitProductResult: of(
      E.left('getAllParentsOfUnitProduct FAILURE'),
    ),
  },
];

test.each(getAllParentsOfUnitProductCases)(
  'getAllParentsOfUnitProduct cases',
  ({ inputState, getAllParentsOfUnitProductResult }, done: any) => {
    const depsFixture = {} as any;

    const funcSpy = jest.fn().mockReturnValue(getAllParentsOfUnitProductResult);
    const spy = jest
      .spyOn(utils, 'getAllParentsOfUnitProduct')
      .mockReturnValue(funcSpy);

    getNextOrderNum(depsFixture)(
      inputState as CalculationState_GroupCurrencyAdded,
    )
      .pipe(
        tap(res => {
          expect(spy.mock.calls).toMatchSnapshot(
            'getAllParentsOfUnitProduct DEPS CALLS',
          );
          expect(funcSpy.mock.calls).toMatchSnapshot(
            'getAllParentsOfUnitProduct FUNC CALLS',
          );
          expect(res).toMatchSnapshot('RESULT');
        }),
      )
      .subscribe(done());
  },
);
