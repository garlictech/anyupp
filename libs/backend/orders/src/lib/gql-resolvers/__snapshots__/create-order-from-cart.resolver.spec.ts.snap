// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOrderFromCart should fail without a cart: GETCART CALLS 1`] = `
Array [
  Array [
    Object {
      "id": "CART ID",
    },
  ],
]
`;

exports[`createOrderFromCart should fail without a cart: GETCART CALLS 2`] = `
Array [
  Array [
    Object {
      "id": "CART ID",
    },
  ],
]
`;

exports[`createOrderFromCart should fail without a cart: GETCART CALLS 3`] = `
Array [
  Array [
    Object {
      "id": "CART ID",
    },
  ],
]
`;

exports[`createOrderFromCart should fail without a cart: RESULT 1`] = `
Object {
  "_tag": "Left",
  "left": "Cart CART ID is missing",
}
`;

exports[`createOrderFromCart should fail without a cart: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "CART ERROR",
}
`;

exports[`createOrderFromCart should fail without a cart: RESULT 3`] = `
Object {
  "_tag": "Left",
  "left": "User ID-s mismatch (USER ID <-> DIFFERENT USER ID)",
}
`;

exports[`getGroupCurrency cases: GETGROUPCURRENCY CALLS 1`] = `
Array [
  Array [
    Object {
      "id": "GROUP ID 1",
    },
  ],
]
`;

exports[`getGroupCurrency cases: GETGROUPCURRENCY CALLS 2`] = `
Array [
  Array [
    Object {
      "id": "GROUP ID 2",
    },
  ],
]
`;

exports[`getGroupCurrency cases: GETGROUPCURRENCY CALLS 3`] = `
Array [
  Array [
    Object {
      "id": "GROUP ID 3",
    },
  ],
]
`;

exports[`getGroupCurrency cases: RESULT 1`] = `
Object {
  "_tag": "Left",
  "left": "Group GROUP ID 1 cannot be fetched from the database.",
}
`;

exports[`getGroupCurrency cases: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "GROUP CURRENCY FETCH FAILURE",
}
`;

exports[`getGroupCurrency cases: RESULT 3`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {
      "id": "CART ID 3",
    },
    "groupCurrency": "GROUP CURRENCY",
    "unit": Object {
      "groupId": "GROUP ID 3",
    },
  },
}
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 1`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 2`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 3`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM DEPS CALLS 4`] = `
Array [
  Array [
    "UNIT TABLE NAME",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 1`] = `
Array [
  Array [
    "UNIT ID 1",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 2`] = `
Array [
  Array [
    "UNIT ID 2",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 3`] = `
Array [
  Array [
    "UNIT ID 3",
  ],
]
`;

exports[`getNextOrderNum cases: INCREMENTORDERNUM FUNC CALLS 4`] = `
Array [
  Array [
    "UNIT ID 4",
  ],
]
`;

exports[`getNextOrderNum cases: RESULT 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {
      "place": Object {
        "seat": "SEAT1",
        "table": "TABLE1",
      },
    },
    "orderNum": "TABLE1SEAT105",
    "unit": Object {
      "id": "UNIT ID 1",
    },
  },
}
`;

exports[`getNextOrderNum cases: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "INCREMENT ORDER NUM FAILURE",
}
`;

exports[`getNextOrderNum cases: RESULT 3`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {
      "place": Object {
        "seat": "SEAT3",
        "table": "TABLE3",
      },
    },
    "orderNum": "TABLE3SEAT310",
    "unit": Object {
      "id": "UNIT ID 3",
    },
  },
}
`;

exports[`getNextOrderNum cases: RESULT 4`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {},
    "orderNum": "20",
    "unit": Object {
      "id": "UNIT ID 4",
    },
  },
}
`;

exports[`getUnit cases: GETUNIT CALLS 1`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 2`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 3`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 4`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 5`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 6`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: GETUNIT CALLS 7`] = `
Array [
  Array [
    Object {
      "id": "UNIT ID",
    },
  ],
]
`;

exports[`getUnit cases: RESULT 1`] = `
Object {
  "_tag": "Left",
  "left": "Unit UNIT ID in the cart cannot be fetched from the database.",
}
`;

exports[`getUnit cases: RESULT 2`] = `
Object {
  "_tag": "Left",
  "left": "UNIT ERROR",
}
`;

exports[`getUnit cases: RESULT 3`] = `
Object {
  "_tag": "Left",
  "left": "Unit does not accept orders",
}
`;

exports[`getUnit cases: RESULT 4`] = `
Object {
  "_tag": "Left",
  "left": "Payment mode is not provided in the cart, and the unit does not accept afterpay.",
}
`;

exports[`getUnit cases: RESULT 5`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {
      "unitId": "UNIT ID",
    },
    "unit": Object {
      "id": "UNIT ID 3",
      "isAcceptingOrders": true,
      "orderPaymentPolicy": "afterpay",
    },
  },
}
`;

exports[`getUnit cases: RESULT 6`] = `
Object {
  "_tag": "Left",
  "left": "Payment mode is not provided in the cart, and the unit does not accept afterpay.",
}
`;

exports[`getUnit cases: RESULT 7`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "cart": Object {
      "paymentMode": Object {},
      "unitId": "UNIT ID",
    },
    "unit": Object {
      "id": "UNIT ID 5",
      "isAcceptingOrders": true,
    },
  },
}
`;
